_posts/2023-04-19-more-optimizations.md

```diff
@@ -0,0 +1,1300 @@
+    jne label_3
+# read 32
+    mov rdi, qword ptr [rsi+14]
+    add qword ptr [rsi+22], 32
+    mov rax, rcx
+    shr rax, 3
+    add rdi, rax
+    and ecx, 7
+    jnz L38
+    movbe edx, dword ptr [rdi]
+    add ecx, 32
+    short jmp L40
+L38:
+    mov rdx, qword ptr [rdi-3]
+    shr rdx, 24
+    bswap rdx
+L40:
+    shl rdx, cl
+# extract integer 8
+    mov rax, rdx
+# store extracted integer as a small
+    shr rax, 52
+    or rax, 15
+    mov qword ptr [rbx+16], rax
+    shl rdx, 8
+# extract integer 24
+    shr rdx, 36
+    or rdx, 15
+    mov qword ptr [rbx+24], rdx
+```
+
+The first part of the code ensures that there are exactly 32 bits
+remaining in the binary:
+
+
+```nasm
+# ensure_exactly 32
+    mov rsi, qword ptr [rbx+8]    ; Get pointer to match context
+    mov rax, qword ptr [rsi+30]   ; Get size of binary in bits
+    mov rcx, qword ptr [rsi+22]   ; Get offset in bits into binary
+    sub rax, rcx
+    cmp rax, 32
+    jne label_3
+```
+
+The next part of the code does not directly correspond to the commands
+in the `bs_match` BEAM instruction. Instead, the code reads 32 bits
+from the binary:
+
+```nasm
+# read 32
+    mov rdi, qword ptr [rsi+14]
+    add qword ptr [rsi+22], 32  ; Increment bit offset in match context
+    mov rax, rcx
+    shr rax, 3
+    add rdi, rax
+    and ecx, 7                  ; Test alignment
+    jnz L38                     ; Jump if segment not byte-aligned
+
+    ; Read 32 bits (4 bytes) byte-aligned and convert to big-endian
+    movbe edx, dword ptr [rdi]
+    add ecx, 32
+    short jmp L40
+
+L38:
+    ; Read a 8-byte word and extract the 32 bits that are needed.
+    mov rdx, qword ptr [rdi-3]
+    shr rdx, 24
+    bswap rdx                   ; Convert to big-endian
+
+L40:
+    ; Shift the read bytes to the most significant bytes of the word
+    shl rdx, cl
+```
+
+The 4 bytes read will be converted to big-endian and placed as the
+most significant bytes of CPU register `rdx` with the rest of the
+register zeroed.
+
+The following instructions extracts the 8 bits for the first segment and
+stores it as a tagged integer in `{x,2}`:
+
+```nasm
+# extract integer 8
+    mov rax, rdx
+# store extracted integer as a small
+    shr rax, 52
+    or rax, 15
+    mov qword ptr [rbx+16], rax
+    shl rdx, 8
+```
+
+The following instructions extracts the 24 bits for the second segment and
+stores it as a tagged integer in `{x,3}`:
+
+```nasm
+# extract integer 24
+    shr rdx, 36
+    or rdx, 15
+    mov qword ptr [rbx+24], rdx
+```
+
+### Binary construction in OTP 26
+
+For binary construction in OTP 26, the compiler emits a
+`bs_create_bin` BEAM instruction just as in OTP 25. However, the
+native code that the JIT in OTP 26 emits for that instruction bears
+little resemblance to the native code emitted by OTP 25. The native
+code will do the following:
+
+* Allocate room on the heap for a binary and initialize it with
+inlined native code. A helper fragment is called if there is not
```

Comment:
I find the term "helper fragment"  unclear.
"helper code fragment"?
"allocation helper code fragment"?
"heap fragment allocation helper function"?
"heap fragment allocation helper code fragment"?

Reply:
I have pushed https://github.com/bjorng/erlang-org/pull/2/commits/6a4e689e1175bc081e28585cb8e0f46cfd751d63, which will use "helper code fragment" consistently and I have also added a short explanation at the place of its first use.